saw-core/src/SAWCore/Term/CtxTerm.hs

```diff
@@ -369,232 +107,143 @@ instance (MonadTerm m, MonadTrans t, Monad (t m)) => MonadTerm (t m) where
 mkFlatTermF :: MonadTerm m => FlatTermF Term -> m Term
 mkFlatTermF = mkTermF . FTermF
 
--- | Build a free variable as a 'CtxTerm'
-ctxVar :: MonadTerm m => Bindings tp (ctx1 ::> a) ctx2 ->
-          m (CtxTerm (CtxInvApp (ctx1 ::> a) ctx2) a)
-ctxVar ctx = CtxTerm <$> mkTermF (LocalVar $ bindingsLength ctx)
+-- | Build a free variable as a 'Term'
+ctxVar :: MonadTerm m => [(LocalName, tp)] -> m Term
+ctxVar ctx = mkTermF (LocalVar $ length ctx)
 
--- | Build a list of all the free variables as 'CtxTerm's
---
--- FIXME: there should be a nicer way to do this that does not require
--- ctxAppNilEq
-ctxVars :: MonadTerm m => InvBindings tp EmptyCtx ctx -> m (CtxTermsCtx ctx ctx)
+-- | Build a list of all the free variables as 'Term's
+ctxVars :: MonadTerm m => [(LocalName, tp)] -> m [Term]
 ctxVars ctx_top =
-  case ctxAppNilEq ctx_top of
-    Refl -> helper ctx_top NoBind
+  helper ctx_top []
       where
-        helper :: MonadTerm m => InvBindings tp EmptyCtx ctx ->
-                  Bindings tp (EmptyCtx <+> ctx) as ->
-                  m (CtxTermsCtx (CtxInvApp (EmptyCtx <+> ctx) as) ctx)
-        helper InvNoBind _ = return CtxTermsCtxNil
-        helper (InvBind vars_ctx x tp) ctx =
-          CtxTermsCtxCons <$> helper vars_ctx (Bind x tp ctx) <*> ctxVar ctx
+        helper :: MonadTerm m => [(LocalName, tp)] -> [(LocalName, tp)] -> m [Term]
+        helper [] _ = return []
+        helper vars_ctx ctx =
+          snoc <$> helper (init vars_ctx) (last vars_ctx : ctx) <*> ctxVar ctx
+        snoc xs x = xs ++ [x]
```

Comment:
Now that we are no longer using `InvBindings`, the time complexity of snoc'ing an additional element has grown from O(1) to O(n). I wonder if instead of replacing `InvBindings` with lists, we should introduce a `SnocList` data list with an O(1) `snoc` operation to maintain the same time complexity.

Reply:
Fixed in a06b2d4841f85a4e7a516be7d6e407480dee2103.